crates/e2e/tests/e2e/wrapper.rs

```diff
@@ -0,0 +1,256 @@
+        std::env::var("FORK_URL_MAINNET")
+            .expect("FORK_URL_MAINNET must be set to run forked tests"),
+        FORK_BLOCK_MAINNET,
+    )
+    .await;
+}
+
+/// Test that orders can be placed with wrapper contracts specified in the app
+/// data. This replicates the functionality of the playground/test_wrapper.sh
+/// script as a proper E2E test.
+async fn forked_mainnet_wrapper_test(web3: Web3) {
+    let mut onchain = OnchainComponents::deployed(web3.clone()).await;
+
+    let [solver] = onchain.make_solvers_forked(to_wei(1)).await;
+    let [trader] = onchain.make_accounts(to_wei(2)).await;
+
+    let token_weth = onchain.contracts().weth.clone();
+    let token_usdc = ERC20::Instance::new(
+        address!("a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"),
+        web3.alloy.clone(),
+    );
+
+    // Authorize the empty wrapper as a solver
+    web3.alloy
+        .anvil_send_impersonated_transaction_with_config(
+            onchain
+                .contracts()
+                .gp_authenticator
+                .addSolver(EMPTY_WRAPPER_MAINNET)
+                .from(
+                    onchain
+                        .contracts()
+                        .gp_authenticator
+                        .manager()
+                        .call()
+                        .await
+                        .unwrap(),
+                )
+                .into_transaction_request(),
+            ImpersonateConfig {
+                fund_amount: Some(to_wei(1).into_alloy()),
+                stop_impersonate: true,
+            },
+        )
+        .await
+        .unwrap()
+        .get_receipt()
+        .await
+        .unwrap();
+
+    // Trader deposits ETH to get WETH
+    web3.alloy
+        .anvil_send_impersonated_transaction_with_config(
+            token_weth
+                .deposit()
+                .value(to_wei(1).into_alloy())
+                .from(trader.address().into_alloy())
+                .into_transaction_request(),
+            ImpersonateConfig {
+                fund_amount: None,
+                stop_impersonate: true,
+            },
+        )
+        .await
+        .unwrap()
+        .get_receipt()
+        .await
+        .unwrap();
+
+    // Approve GPv2 for trading
+    token_weth
+        .approve(
+            onchain.contracts().allowance.into_alloy(),
+            to_wei(1).into_alloy(),
+        )
+        .from(trader.address().into_alloy())
+        .send_and_watch()
+        .await
+        .unwrap();
+
+    // Start services
+    let services = Services::new(&onchain).await;
+    services.start_protocol(solver).await;
+
+    onchain.mint_block().await;
+
+    // Create app data with the deployed EmptyWrapper contract
+    let app_data = json!({
+        "version": "0.9.0",
+        "metadata": {
+            "wrappers": [{
+                "address": format!("{:?}", EMPTY_WRAPPER_MAINNET),
+                "data": "0xbeef",
+                "isOmittable": false,
```

Comment:
nit: Given that wrappers are composable and a lot of debate went into encoding them correctly could you check if it's easy to have 2 wrappers here and use `trace_transaction` to verify that 1 wrapper indeed called the other wrapper. If it's too much work I think it would still be slightly better test coverage to have 2 wrappers here and still only assert that `tx.to` was the wrapper contract.